File: learning-kmp/bookshelf/iosApp/iosApp/DesignSystem/BookshelfIcons.swift
import SwiftUI

struct PencilIcon: View {
    var color: Color = .primary
    var size: CGFloat = 24
    
    var body: some View {
        Canvas { ctx, sz in
            let w = sz.width
            let h = sz.height
            
            // Scale stroke width based on size, with a minimum floor
            let lw = max(1.5, w * 0.08)
            let style = StrokeStyle(lineWidth: lw, lineCap: .round, lineJoin: .round)
            
            var pencil = Path()
            
            // Main Outline
            pencil.move(to: CGPoint(x: w * 0.125, y: h * 0.875)) // Bottom-left tip
            pencil.addLine(to: CGPoint(x: w * 0.250, y: h * 0.625)) // Left tip corner
            pencil.addLine(to: CGPoint(x: w * 0.625, y: h * 0.250)) // Left eraser base
            pencil.addLine(to: CGPoint(x: w * 0.750, y: h * 0.125)) // Top-left eraser
            pencil.addLine(to: CGPoint(x: w * 0.875, y: h * 0.250)) // Top-right eraser
            pencil.addLine(to: CGPoint(x: w * 0.750, y: h * 0.375)) // Right eraser base
            pencil.addLine(to: CGPoint(x: w * 0.375, y: h * 0.750)) // Right tip corner
            pencil.closeSubpath()
            
            // Eraser separator line
            pencil.move(to: CGPoint(x: w * 0.625, y: h * 0.250))
            pencil.addLine(to: CGPoint(x: w * 0.750, y: h * 0.375))
            
            // Tip separator line
            pencil.move(to: CGPoint(x: w * 0.250, y: h * 0.625))
            pencil.addLine(to: CGPoint(x: w * 0.375, y: h * 0.750))
            
            ctx.stroke(pencil, with: .color(color), style: style)
        }
        .frame(width: size, height: size)
    }
}

struct UndoIcon: View {
    var color: Color = .primary
    var size: CGFloat = 24
    var body: some View {
        Canvas { ctx, sz in
            let w = sz.width, h = sz.height
            let lw = max(1.5, w * 0.10)
            let style = StrokeStyle(lineWidth: lw, lineCap: .round, lineJoin: .round)
            var arrow = Path()
            arrow.move(to: CGPoint(x: w*0.35, y: h*0.28))
            arrow.addLine(to: CGPoint(x: w*0.18, y: h*0.44))
            arrow.addLine(to: CGPoint(x: w*0.35, y: h*0.60))
            var arc = Path()
            arc.move(to: CGPoint(x: w*0.18, y: h*0.44))
            arc.addLine(to: CGPoint(x: w*0.68, y: h*0.44))
            arc.addCurve(
                to: CGPoint(x: w*0.68, y: h*0.74),
                control1: CGPoint(x: w*0.85, y: h*0.44),
                control2: CGPoint(x: w*0.85, y: h*0.74)
            )
            arc.addLine(to: CGPoint(x: w*0.40, y: h*0.74))
            ctx.stroke(arrow, with: .color(color), style: style)
            ctx.stroke(arc, with: .color(color), style: style)
        }
        .frame(width: size, height: size)
    }
}

struct BookIcon: View {
    var color: Color = .primary
    var size: CGFloat = 24
    var body: some View {
        Canvas { ctx, sz in
            let w = sz.width, h = sz.height
            let lw = max(1.5, w * 0.065)
            var left = Path()
            left.move(to: CGPoint(x: w*0.5, y: h*0.14))
            left.addLine(to: CGPoint(x: w*0.06, y: h*0.24))
            left.addLine(to: CGPoint(x: w*0.06, y: h*0.84))
            left.addLine(to: CGPoint(x: w*0.5, y: h*0.76))
            left.closeSubpath()
            var right = Path()
            right.move(to: CGPoint(x: w*0.5, y: h*0.14))
            right.addLine(to: CGPoint(x: w*0.94, y: h*0.24))
            right.addLine(to: CGPoint(x: w*0.94, y: h*0.84))
            right.addLine(to: CGPoint(x: w*0.5, y: h*0.76))
            right.closeSubpath()
            var spine = Path()
            spine.move(to: CGPoint(x: w*0.5, y: h*0.14))
            spine.addLine(to: CGPoint(x: w*0.5, y: h*0.76))
            let style = StrokeStyle(lineWidth: lw, lineCap: .round, lineJoin: .round)
            ctx.stroke(left, with: .color(color), style: style)
            ctx.stroke(right, with: .color(color), style: style)
            ctx.stroke(spine, with: .color(color), style: StrokeStyle(lineWidth: lw * 0.6, lineCap: .round))
        }
        .frame(width: size, height: size)
    }
}

struct BooksStackIcon: View {
    var color: Color = .primary
    var size: CGFloat = 24
    var body: some View {
        Canvas { ctx, sz in
            let w = sz.width, h = sz.height
            let lw = max(1.5, w * 0.08)
            let bookW = w * 0.20
            let gap = w * 0.075
            let baseY = h * 0.88
            let configs: [(x: CGFloat, topY: CGFloat)] = [
                (w*0.08, h*0.18),
                (w*0.08 + bookW + gap, h*0.30),
                (w*0.08 + (bookW+gap)*2, h*0.22)
            ]
            for cfg in configs {
                var book = Path()
                book.addRoundedRect(
                    in: CGRect(x: cfg.x, y: cfg.topY, width: bookW, height: baseY - cfg.topY),
                    cornerSize: CGSize(width: 2, height: 2)
                )
                ctx.stroke(book, with: .color(color), style: StrokeStyle(lineWidth: lw, lineCap: .round, lineJoin: .round))
            }
            var base = Path()
            base.move(to: CGPoint(x: w*0.04, y: baseY))
            base.addLine(to: CGPoint(x: w*0.96, y: baseY))
            ctx.stroke(base, with: .color(color), style: StrokeStyle(lineWidth: lw, lineCap: .round))
        }
        .frame(width: size, height: size)
    }
}

struct PlusIcon: View {
    var color: Color = .primary
    var size: CGFloat = 24
    var body: some View {
        Canvas { ctx, sz in
            let w = sz.width, h = sz.height
            let lw = max(2, w * 0.13)
            let style = StrokeStyle(lineWidth: lw, lineCap: .round)
            var h_line = Path()
            h_line.move(to: CGPoint(x: w*0.22, y: h*0.5))
            h_line.addLine(to: CGPoint(x: w*0.78, y: h*0.5))
            var v_line = Path()
            v_line.move(to: CGPoint(x: w*0.5, y: h*0.22))
            v_line.addLine(to: CGPoint(x: w*0.5, y: h*0.78))
            ctx.stroke(h_line, with: .color(color), style: style)
            ctx.stroke(v_line, with: .color(color), style: style)
        }
        .frame(width: size, height: size)
    }
}

struct TrashIcon: View {
    var color: Color = .primary
    var size: CGFloat = 24
    var body: some View {
        Canvas { ctx, sz in
            let w = sz.width, h = sz.height
            let lw = max(1.5, w * 0.09)
            let style = StrokeStyle(lineWidth: lw, lineCap: .round, lineJoin: .round)
            var bin = Path()
            bin.move(to: CGPoint(x: w*0.20, y: h*0.30))
            bin.addLine(to: CGPoint(x: w*0.26, y: h*0.87))
            bin.addLine(to: CGPoint(x: w*0.74, y: h*0.87))
            bin.addLine(to: CGPoint(x: w*0.80, y: h*0.30))
            var lid = Path()
            lid.move(to: CGPoint(x: w*0.11, y: h*0.30))
            lid.addLine(to: CGPoint(x: w*0.89, y: h*0.30))
            var handle = Path()
            handle.move(to: CGPoint(x: w*0.38, y: h*0.30))
            handle.addLine(to: CGPoint(x: w*0.38, y: h*0.17))
            handle.addLine(to: CGPoint(x: w*0.62, y: h*0.17))
            handle.addLine(to: CGPoint(x: w*0.62, y: h*0.30))
            ctx.stroke(bin, with: .color(color), style: style)
            ctx.stroke(lid, with: .color(color), style: StrokeStyle(lineWidth: lw, lineCap: .round))
            ctx.stroke(handle, with: .color(color), style: style)
        }
        .frame(width: size, height: size)
    }
}

struct ChevronLeftIcon: View {
    var color: Color = .primary
    var size: CGFloat = 24
    var body: some View {
        Canvas { ctx, sz in
            let w = sz.width, h = sz.height
            var path = Path()
            path.move(to: CGPoint(x: w*0.63, y: h*0.22))
            path.addLine(to: CGPoint(x: w*0.37, y: h*0.50))
            path.addLine(to: CGPoint(x: w*0.63, y: h*0.78))
            ctx.stroke(path, with: .color(color), style: StrokeStyle(lineWidth: max(2, w*0.11), lineCap: .round, lineJoin: .round))
        }
        .frame(width: size, height: size)
    }
}

struct ChevronRightIcon: View {
    var color: Color = .primary
    var size: CGFloat = 24
    var body: some View {
        Canvas { ctx, sz in
            let w = sz.width, h = sz.height
            var path = Path()
            path.move(to: CGPoint(x: w*0.37, y: h*0.22))
            path.addLine(to: CGPoint(x: w*0.63, y: h*0.50))
            path.addLine(to: CGPoint(x: w*0.37, y: h*0.78))
            ctx.stroke(path, with: .color(color), style: StrokeStyle(lineWidth: max(2, w*0.11), lineCap: .round, lineJoin: .round))
        }
        .frame(width: size, height: size)
    }
}

File: learning-kmp/bookshelf/androidApp/src/main/kotlin/com/bookshelf/android/ui/BookshelfIcons.kt
package com.bookshelf.android.ui

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.size
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.RoundRect
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.StrokeJoin
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import kotlin.math.max

@Composable
fun BookIcon(
    contentDescription: String? = null,
    tint: Color = Color.Unspecified,
    size: Dp = 24.dp,
    modifier: Modifier = Modifier
) {
    Canvas(modifier = modifier.size(size).maybeSemantics(contentDescription)) {
        val w = this.size.width; val h = this.size.height
        val lw = max(1.5f, w * 0.065f)
        val stroke = Stroke(lw, cap = StrokeCap.Round, join = StrokeJoin.Round)
        val c = if (tint == Color.Unspecified) Color(0xFF49454F) else tint
        drawPath(Path().apply {
            moveTo(w * .5f, h * .14f); lineTo(w * .06f, h * .24f)
            lineTo(w * .06f, h * .84f); lineTo(w * .5f, h * .76f); close()
        }, c, style = stroke)
        drawPath(Path().apply {
            moveTo(w * .5f, h * .14f); lineTo(w * .94f, h * .24f)
            lineTo(w * .94f, h * .84f); lineTo(w * .5f, h * .76f); close()
        }, c, style = stroke)
        drawLine(c, Offset(w * .5f, h * .14f), Offset(w * .5f, h * .76f),
            strokeWidth = max(1f, lw * .6f), cap = StrokeCap.Round)
    }
}

@Composable
fun BooksStackIcon(
    contentDescription: String? = null,
    tint: Color = Color.Unspecified,
    size: Dp = 24.dp,
    modifier: Modifier = Modifier
) {
    Canvas(modifier = modifier.size(size).maybeSemantics(contentDescription)) {
        val w = this.size.width; val h = this.size.height
        val lw = max(1.5f, w * 0.08f)
        val stroke = Stroke(lw, cap = StrokeCap.Round, join = StrokeJoin.Round)
        val c = if (tint == Color.Unspecified) Color(0xFF49454F) else tint
        val bw = w * .20f; val gap = w * .075f; val baseY = h * .88f
        val configs = listOf(
            Pair(w * .08f, h * .18f),
            Pair(w * .08f + bw + gap, h * .30f),
            Pair(w * .08f + (bw + gap) * 2, h * .22f)
        )
        for ((x, topY) in configs) {
            drawPath(Path().apply {
                addRoundRect(RoundRect(x, topY, x + bw, baseY, CornerRadius(2f)))
            }, c, style = stroke)
        }
        drawLine(c, Offset(w * .04f, baseY), Offset(w * .96f, baseY),
            strokeWidth = lw, cap = StrokeCap.Round)
    }
}

@Composable
fun PlusIcon(
    contentDescription: String? = null,
    tint: Color = Color.Unspecified,
    size: Dp = 24.dp,
    modifier: Modifier = Modifier
) {
    Canvas(modifier = modifier.size(size).maybeSemantics(contentDescription)) {
        val w = this.size.width; val h = this.size.height
        val lw = max(2f, w * .13f)
        val c = if (tint == Color.Unspecified) Color(0xFF49454F) else tint
        drawLine(c, Offset(w * .22f, h * .5f), Offset(w * .78f, h * .5f), lw, StrokeCap.Round)
        drawLine(c, Offset(w * .5f, h * .22f), Offset(w * .5f, h * .78f), lw, StrokeCap.Round)
    }
}

@Composable
fun TrashIcon(
    contentDescription: String? = null,
    tint: Color = Color.Unspecified,
    size: Dp = 24.dp,
    modifier: Modifier = Modifier
) {
    Canvas(modifier = modifier.size(size).maybeSemantics(contentDescription)) {
        val w = this.size.width; val h = this.size.height
        val lw = max(1.5f, w * .09f)
        val stroke = Stroke(lw, cap = StrokeCap.Round, join = StrokeJoin.Round)
        val c = if (tint == Color.Unspecified) Color(0xFF49454F) else tint
        drawPath(Path().apply {
            moveTo(w * .20f, h * .30f); lineTo(w * .26f, h * .87f)
            lineTo(w * .74f, h * .87f); lineTo(w * .80f, h * .30f)
        }, c, style = stroke)
        drawLine(c, Offset(w * .11f, h * .30f), Offset(w * .89f, h * .30f), lw, StrokeCap.Round)
        drawPath(Path().apply {
            moveTo(w * .38f, h * .30f); lineTo(w * .38f, h * .17f)
            lineTo(w * .62f, h * .17f); lineTo(w * .62f, h * .30f)
        }, c, style = stroke)
    }
}

@Composable
fun BackArrowIcon(
    contentDescription: String? = null,
    tint: Color = Color.Unspecified,
    size: Dp = 24.dp,
    modifier: Modifier = Modifier
) {
    Canvas(modifier = modifier.size(size).maybeSemantics(contentDescription)) {
        val w = this.size.width; val h = this.size.height
        val lw = max(2f, w * .10f)
        val stroke = Stroke(lw, cap = StrokeCap.Round, join = StrokeJoin.Round)
        val c = if (tint == Color.Unspecified) Color(0xFF49454F) else tint
        drawPath(Path().apply {
            moveTo(w * .58f, h * .22f); lineTo(w * .26f, h * .50f); lineTo(w * .58f, h * .78f)
        }, c, style = stroke)
        drawLine(c, Offset(w * .26f, h * .50f), Offset(w * .82f, h * .50f), lw, StrokeCap.Round)
    }
}

@Composable
fun ChevronLeftIcon(
    contentDescription: String? = null,
    tint: Color = Color.Unspecified,
    size: Dp = 24.dp,
    modifier: Modifier = Modifier
) {
    Canvas(modifier = modifier.size(size).maybeSemantics(contentDescription)) {
        val w = this.size.width; val h = this.size.height
        val lw = max(2f, w * .11f)
        val c = if (tint == Color.Unspecified) Color(0xFF49454F) else tint
        drawPath(Path().apply {
            moveTo(w * .63f, h * .22f); lineTo(w * .37f, h * .50f); lineTo(w * .63f, h * .78f)
        }, c, style = Stroke(lw, cap = StrokeCap.Round, join = StrokeJoin.Round))
    }
}

@Composable
fun ChevronRightIcon(
    contentDescription: String? = null,
    tint: Color = Color.Unspecified,
    size: Dp = 24.dp,
    modifier: Modifier = Modifier
) {
    Canvas(modifier = modifier.size(size).maybeSemantics(contentDescription)) {
        val w = this.size.width; val h = this.size.height
        val lw = max(2f, w * .11f)
        val c = if (tint == Color.Unspecified) Color(0xFF49454F) else tint
        drawPath(Path().apply {
            moveTo(w * .37f, h * .22f); lineTo(w * .63f, h * .50f); lineTo(w * .37f, h * .78f)
        }, c, style = Stroke(lw, cap = StrokeCap.Round, join = StrokeJoin.Round))
    }
}

@Composable
fun PencilIcon(
    contentDescription: String? = null,
    tint: Color = Color.Unspecified,
    size: Dp = 24.dp,
    modifier: Modifier = Modifier
) {
    val semanticsModifier = if (contentDescription != null) {
        Modifier.semantics { this.contentDescription = contentDescription }
    } else Modifier

    Canvas(modifier = modifier.size(size).then(semanticsModifier)) {
        val w = this.size.width
        val h = this.size.height
        
        // Scale stroke width based on size, with a minimum floor
        val lw = max(1.5f, w * 0.08f)
        val stroke = Stroke(width = lw, cap = StrokeCap.Round, join = StrokeJoin.Round)
        val c = if (tint == Color.Unspecified) Color(0xFF49454F) else tint

        val pencilPath = Path().apply {
            // Main Outline
            moveTo(w * 0.125f, h * 0.875f) // Bottom-left tip
            lineTo(w * 0.250f, h * 0.625f) // Left tip corner
            lineTo(w * 0.625f, h * 0.250f) // Left eraser base
            lineTo(w * 0.750f, h * 0.125f) // Top-left eraser
            lineTo(w * 0.875f, h * 0.250f) // Top-right eraser
            lineTo(w * 0.750f, h * 0.375f) // Right eraser base
            lineTo(w * 0.375f, h * 0.750f) // Right tip corner
            close()

            // Eraser separator line
            moveTo(w * 0.625f, h * 0.250f)
            lineTo(w * 0.750f, h * 0.375f)

            // Tip separator line
            moveTo(w * 0.250f, h * 0.625f)
            lineTo(w * 0.375f, h * 0.750f)
        }

        drawPath(pencilPath, color = c, style = stroke)
    }
}

@Composable
fun UndoIcon(
    contentDescription: String? = null,
    tint: Color = Color.Unspecified,
    size: Dp = 24.dp,
    modifier: Modifier = Modifier
) {
    Canvas(modifier = modifier.size(size).maybeSemantics(contentDescription)) {
        val w = this.size.width; val h = this.size.height
        val lw = max(1.5f, w * 0.10f)
        val c = if (tint == Color.Unspecified) Color(0xFF49454F) else tint
        drawPath(Path().apply {
            moveTo(w * .35f, h * .28f); lineTo(w * .18f, h * .44f); lineTo(w * .35f, h * .60f)
        }, c, style = Stroke(lw, cap = StrokeCap.Round, join = StrokeJoin.Round))
        drawPath(Path().apply {
            moveTo(w * .18f, h * .44f); lineTo(w * .68f, h * .44f)
            cubicTo(w * .85f, h * .44f, w * .85f, h * .74f, w * .68f, h * .74f)
            lineTo(w * .40f, h * .74f)
        }, c, style = Stroke(lw, cap = StrokeCap.Round, join = StrokeJoin.Round))
    }
}

private fun Modifier.maybeSemantics(description: String?): Modifier =
    if (description != null) this.semantics { contentDescription = description } else this
